### PR TITLE
feat: add methods to managers

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**

This PR adds a bunch of methods to managers (without any intention to remove them from the structures of course). For some of these I had to use resolve instead of resolveId due to how the code works so if you believe this defeats the purpose of a manager method I can try refactoring or undoing some of those changes. This PR sort of continues the work from #7286 

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 1]
```diff
@@ -170,6 +173,149 @@ class GuildChannelManager extends CachedManager {
...
+
+  /**
+   * Edits the channel.
+   * @param {GuildChannelResolvable} channel The channel to edit
+   * @param {ChannelData} data The new data for the channel
+   * @param {string} [reason] Reason for editing this channel
+   * @returns {Promise<GuildChannel>}
+   * @example
+   * // Edit a channel
+   * channel.edit({ name: 'new-channel' })
```
Comment (imranbarbhuiya) [inline packages/discord.js/src/managers/GuildChannelManager.js:236]:
```suggestion
   * guild.channels.edit('222197033908436994', { name: 'new-channel' })
```

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 2]
```diff
@@ -260,6 +425,23 @@ class GuildChannelManager extends CachedManager {
...
   }
+
+  /**
+   * Deletes the channel.
+   * @param {GuildChannelResolvable} channel The channel to delete
+   * @param {string} [reason] Reason for deleting this channel
+   * @returns {Promise<void>}
+   * @example
+   * // Delete the channel
+   * channel.delete('858850993013260338', 'making room for new channels')
```
Comment (imranbarbhuiya) [inline packages/discord.js/src/managers/GuildChannelManager.js:436]:
```suggestion
   * guild.channels.delete('858850993013260338', 'making room for new channels')
```

File: packages/discord.js/src/managers/RoleManager.js [hunk 3]
```diff
@@ -224,7 +225,37 @@ class RoleManager extends CachedManager {
...
-  /*
+  /**
+   * Sets the new position of the role.
+   * @param {RoleResolvable} role The role to change the position of
+   * @param {number} position The new position for the role
+   * @param {SetRolePositionOptions} [options] Options for setting the position
+   * @returns {Promise<Role>}
+   * @example
+   * // Set the position of the role
+   * role.setPosition(1)
```
Comment (imranbarbhuiya) [inline packages/discord.js/src/managers/RoleManager.js:236]:
```suggestion
   * guild.roles.setPosition('222197033908436994', 1)
```

File: packages/discord.js/typings/index.d.ts [hunk 4]
```diff
@@ -3074,6 +3089,7 @@ export class RoleManager extends CachedManager<Snowflake, Role, RoleResolvable>
   public create(options?: CreateRoleOptions): Promise<Role>;
   public edit(role: RoleResolvable, options: RoleData, reason?: string): Promise<Role>;
   public delete(role: RoleResolvable, reason?: string): Promise<void>;
+  public setPosition(role: RoleResolvable, position: number, options: SetRolePositionOptions): Promise<Role>;
```
Comment (imranbarbhuiya) [inline packages/discord.js/typings/index.d.ts:3092]:
```suggestion
  public setPosition(role: RoleResolvable, position: number, options?: SetRolePositionOptions): Promise<Role>;
```

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/discord.js/typings/index.d.ts [hunk 5]
```diff
@@ -2846,10 +2846,23 @@ export class GuildChannelManager extends CachedManager<Snowflake, GuildBasedChan
...
   public create(name: string, options: GuildChannelCreateOptions): Promise<NonThreadGuildBasedChannel>;
+  public createWebhook(
+    channel: GuildChannelResolvable,
+    name: string,
+    options?: ChannelWebhookCreateOptions,
+  ): Promise<Webhook>;
+  public edit(channel: GuildChannelResolvable, data: ChannelData, reason?: string): Promise<GuildChannel>;
   public fetch(id: Snowflake, options?: BaseFetchOptions): Promise<NonThreadGuildBasedChannel | null>;
   public fetch(id?: undefined, options?: BaseFetchOptions): Promise<Collection<Snowflake, NonThreadGuildBasedChannel>>;
+  public fetcWebhooks(channel: GuildChannelResolvable): Promise<Collection<Snowflake, Webhook>>;
```
Comment (suneettipirneni) [inline packages/discord.js/typings/index.d.ts:2857]:
```suggestion
  public fetchWebhooks(channel: GuildChannelResolvable): Promise<Collection<Snowflake, Webhook>>;
```

**Review by suneettipirneni — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
@iCrawl This should be semver minor, not major ?

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 6]
```diff
@@ -169,6 +172,149 @@ class GuildChannelManager extends CachedManager {
...
+   * @example
+   * // Edit a channel
+   * guild.channels.edit('222197033908436994', { name: 'new-channel' })
+   *   .then(console.log)
+   *   .catch(console.error);
+   */
+  async edit(channel, data, reason) {
+    channel = this.resolve(channel);
+    if (!channel) throw new TypeError('INVALID_TYPE', 'channel', 'GuildChannelResolvable');
+    data.parent &&= this.client.channels.resolveId(data.parent);
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:242]: This should be changed to not mutate data, we can use this:
```suggestion

    const parent = data.parent && this.client.channels.resolveId(data.parent);
```

And update the code below accordingly.

Also added a newline on top to separate different logic blocks.

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 7]
```diff
@@ -169,6 +172,149 @@ class GuildChannelManager extends CachedManager {
...
+
+    if (typeof data.position !== 'undefined') await this.setPosition(channel, data.position, { reason });
+
+    let permission_overwrites;
+
+    if (data.permissionOverwrites) {
+      permission_overwrites = data.permissionOverwrites.map(o => PermissionOverwrites.resolve(o, this.guild));
+    }
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:250]: We can refactor this now:
```suggestion
    let permission_overwrites = data.permissionOverwrites?.map(o => PermissionOverwrites.resolve(o, this.guild));
```

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 8]
```diff
@@ -204,6 +350,25 @@ class GuildChannelManager extends CachedManager {
...
+   *   .then(hooks => console.log(`This channel has ${hooks.size} hooks`))
+   *   .catch(console.error);
+   */
+  async fetchWebhooks(channel) {
+    const id = this.resolveId(channel);
+    if (!id) throw new TypeError('INVALID_TYPE', 'channel', 'GuildChannelResolvable');
+    const data = await this.client.api.channels[id].webhooks.get();
+    const hooks = new Collection();
+    for (const hook of data) hooks.set(hook.id, new Webhook(this.client, hook));
+    return hooks;
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:369]: Use Array#reduce pattern:
```suggestion
    return data.reduce((hooks, hook) => hooks.set(hook.id, new Webhook(this.client, hook)), new Collection());
```

File: packages/discord.js/src/managers/GuildEmojiManager.js [hunk 9]
```diff
@@ -142,6 +143,27 @@ class GuildEmojiManager extends BaseGuildEmojiManager {
...
+    emoji = this.resolve(emoji);
+    if (!emoji) throw new TypeError('INVALID_TYPE', 'emoji', 'EmojiResolvable', true);
+    if (emoji.managed) {
+      throw new Error('EMOJI_MANAGED');
+    } else {
+      if (!this.guild.me) throw new Error('GUILD_UNCACHED_ME');
+      if (!this.guild.me.permissions.has(Permissions.FLAGS.MANAGE_EMOJIS_AND_STICKERS)) {
+        throw new Error('MISSING_MANAGE_EMOJIS_AND_STICKERS_PERMISSION', this.guild);
+      }
+    }
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildEmojiManager.js:162]: No `else` after `throw`, under the same rule as useless `else` after `return`:
```suggestion
    if (emoji.managed) {
      throw new Error('EMOJI_MANAGED');
    }

    const { me } = this.guild;
    if (!me) throw new Error('GUILD_UNCACHED_ME');
    if (!me.permissions.has(Permissions.FLAGS.MANAGE_EMOJIS_AND_STICKERS)) {
      throw new Error('MISSING_MANAGE_EMOJIS_AND_STICKERS_PERMISSION', this.guild);
    }
```

Also tweaked the code a little so the `guild.me` getter isn't called twice.

File: packages/discord.js/src/structures/GuildChannel.js [hunk 10]
```diff
@@ -405,55 +336,8 @@ class GuildChannel extends Channel {
...
-   * @typedef {GuildChannelCreateOptions} GuildChannelCloneOptions
-   * @property {string} [name=this.name] Name of the new channel
-   */
-
-  /**
-   * Clones this channel.
-   * @param {GuildChannelCloneOptions} [options] The options for cloning this channel
-   * @returns {Promise<GuildChannel>}
-   */
-  clone(options = {}) {
```
Comment (kyranet) [inline packages/discord.js/src/structures/GuildChannel.js:443]: Why is clone now gone?

**Review by kyranet — CHANGES_REQUESTED** [review]
Furthermore, replace all usages of `client.api` with the new `client.rest`.

File: packages/discord.js/src/structures/GuildChannel.js (hunk 10 above)
Comment (ImRodry) [reply packages/discord.js/src/structures/GuildChannel.js]: Fixed

File: packages/discord.js/src/managers/GuildChannelManager.js (hunk 6 above)
Comment (ImRodry) [reply packages/discord.js/src/managers/GuildChannelManager.js]: Fixed

**Comment by ImRodry** [comment]
Rebased and applied all suggestions!

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 11]
```diff
@@ -169,6 +172,148 @@ class GuildChannelManager extends CachedManager {
...
+   *   .then(console.log)
+   *   .catch(console.error)
+   */
+  async createWebhook(channel, name, { avatar, reason } = {}) {
+    const id = this.resolveId(channel);
+    if (!id) throw new TypeError('INVALID_TYPE', 'channel', 'GuildChannelResolvable');
+    if (typeof avatar === 'string' && !avatar.startsWith('data:')) {
+      avatar = await DataResolver.resolveImage(avatar);
+    }
+    const data = await this.client.rest.post(Routes.channelWebhooks(this.id), {
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:196]:
```suggestion
    const data = await this.client.rest.post(Routes.channelWebhooks(id), {
```

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 12]
```diff
@@ -169,6 +172,148 @@ class GuildChannelManager extends CachedManager {
...
+          );
+        }
+      } else if (channel.parent) {
+        permission_overwrites = this.parent.permissionOverwrites.cache.map(o =>
+          PermissionOverwrites.resolve(o, this.guild),
+        );
+      }
+    }
+
+    const newData = await this.client.rest.patch(Routes.channel(this.id), {
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:264]:
```suggestion
    const newData = await this.client.rest.patch(Routes.channel(channel.id), {
```

File: packages/discord.js/src/managers/GuildChannelManager.js [hunk 13]
```diff
@@ -204,6 +349,23 @@ class GuildChannelManager extends CachedManager {
...
+   * @example
+   * // Fetch webhooks
+   * guild.channels.fetchWebhooks('769862166131245066')
+   *   .then(hooks => console.log(`This channel has ${hooks.size} hooks`))
+   *   .catch(console.error);
+   */
+  async fetchWebhooks(channel) {
+    const id = this.resolveId(channel);
+    if (!id) throw new TypeError('INVALID_TYPE', 'channel', 'GuildChannelResolvable');
+    const data = await this.client.rest.get(Routes.channelWebhooks(this.id));
```
Comment (kyranet) [inline packages/discord.js/src/managers/GuildChannelManager.js:365]:
```suggestion
    const data = await this.client.rest.get(Routes.channelWebhooks(id));
```

**Review by kyranet — APPROVED** [review]
Just 3 more things, then LGTM.

File: packages/discord.js/src/managers/GuildChannelManager.js (hunk 11 above)
Comment (ImRodry) [reply packages/discord.js/src/managers/GuildChannelManager.js:196]: Thought I got all of those, turns out I missed a few 😅 
All fixed now

**Review by iCrawl — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**